src/core/Threads/ThreadPool.cpp

```diff
@@ -1,41 +1,45 @@
 #include "ThreadPool.hpp"
 
-#include "ThreadWorker.hpp"
-#include "ConsoleLog.hpp"
-
 ThreadPool::ThreadPool(unsigned int threadsNumber)
-  : m_jobsPending(0)
-  , m_threads(std::vector<std::thread>(threadsNumber))
-  , m_shutdown(false)
 {
-  for (auto& thread : m_threads)
-  {
-    m_threads.emplace_back(std::thread(ThreadWorker(this)));
-  }
+  m_threads.reserve(threadsNumber);
+  m_tasks.reserve(threadsNumber);
 }
 
 ThreadPool::~ThreadPool()
 {
   Shutdown();
 }
 
-void ThreadPool::Shutdown()
+void ThreadPool::Submit(std::unique_ptr<ThreadTask> &&task)
 {
-  m_shutdown = true;
+  std::unique_ptr<sf::Thread> thread(new sf::Thread(&ThreadTask::Run, task.get()));
+
+  m_tasks.push_back(std::move(task));
+
+  m_threads.push_back(std::move(thread));
+}
 
-  m_condition.notify_all();
+void ThreadPool::Start()
+{
+  for (auto &thread : m_threads)
+  {
+    thread -> launch();
+  }
+}
 
-  for (std::thread &t : m_threads)
+void ThreadPool::Shutdown()
+{
+  for (auto &thread : m_threads)
   {
-    if (t.joinable())
-    {
-      t.join();
-    }
+    thread -> wait();
   }
+
+  m_threads.clear();
+  m_tasks.clear();
 }
 
-void ThreadPool::WaitForFinish()
+bool ThreadPool::IsRunning() const
 {
-  std::unique_lock<std::mutex> lock(m_mutex);
-  m_condition.wait(lock);
+  return std::any_of(m_tasks.begin(), m_tasks.end(), [](auto &task) { return task -> IsFinished() == false; });
```

Comment:
Maybe use std::all_of and check if IsFinished() is equal to true?

Reply:
I suppose that ::IsRunning name is descriptive enough, no need to change the logic. At least I hope so. xD